source/extensions/filters/network/rocketmq_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,248 @@
+                  host_addr);
+        return true;
+      }
+
+    } else {
+      if (!PermHelper::canWrite(perm)) {
+        ENVOY_LOG(trace, "No permission to write current broker, IP: {}", host_addr);
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+uint32_t RouterImpl::hostSelectionRetryCount() const {
+  auto code = static_cast<RequestCode>(active_message_->downstreamRequest()->code());
+  if (code == RequestCode::SEND_MESSAGE || code == RequestCode::SEND_MESSAGE_V2 ||
+      code == RequestCode::POP_MESSAGE) {
+    std::string cluster_name = active_message_->clusterName();
+    Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(cluster_name);
+    auto& filter_metadata = cluster->info()->metadata().filter_metadata();
+    const auto filter_it = filter_metadata.find(NetworkFilterNames::get().RocketmqProxy);
+    const auto& metadata_fields = filter_it->second.fields();
+    ASSERT(metadata_fields.contains(RocketmqConstants::get().BrokersMeta));
+    uint32_t hosts_num =
+        metadata_fields.at(RocketmqConstants::get().BrokersMeta).struct_value().fields_size();
+    return hosts_num > 1 ? hosts_num : 1;
+  }
+  return 1;
+}
+
+void RouterImpl::onUpstreamData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(trace, "Received some data from upstream: {} bytes, end_stream: {}", data.length(),
+            end_stream);
+  if (active_message_->onUpstreamData(data, end_stream, connection_data_)) {
+    reset();
+  }
+}
+
+void RouterImpl::sendRequestToUpstream(ActiveMessage& active_message) {
+  active_message_ = &active_message;
+  int opaque = active_message_->downstreamRequest()->opaque();
+  std::string cluster_name = active_message_->clusterName();
+  if (cluster_name.empty()) {
+    ENVOY_LOG(warn, "Cluster name is unexpected empty. Opaque: {}", opaque);
+    reset();
+    return;
+  }
+
+  Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(cluster_name);
+  if (!cluster) {
+    ENVOY_LOG(warn, "Cluster for {} is not available", cluster_name);
+    reset();
+    return;
+  }
+
+  cluster_info_ = cluster->info();
+  if (cluster_info_->maintenanceMode()) {
+    ENVOY_LOG(warn, "Cluster {} is under maintenance. Opaque: {}", cluster_name, opaque);
+    active_message.onError("Cluster under maintenance.");
+    active_message.connectionManager().stats().maintenance_failure_.inc();
+    reset();
+    return;
+  }
+
+  Tcp::ConnectionPool::Instance* connectionPool = cluster_manager_.tcpConnPoolForCluster(
+      cluster_name, Upstream::ResourcePriority::Default, this);
+  if (!connectionPool) {
+    ENVOY_LOG(warn, "No host available for cluster {}. Opaque: {}", cluster_name, opaque);
+    active_message.onError("No host available");
+    reset();
+    return;
+  }
+
+  upstream_request_ = std::make_unique<UpstreamRequest>(*this, *connectionPool);
+  Tcp::ConnectionPool::Cancellable* cancellable = connectionPool->newConnection(*upstream_request_);
```

Comment:
You need to cancel the connection in `RouterImpl` destructor

Reply:
OK.